### ISSUE TITLE
Mousepointer does not respond till idle time

### ISSUE THREAD

**Issue by dmrvb** [twinbasic/twinbasic#1329]
**Describe the bug**
Setting Mousepointer does NOT respond till the system becomes idle.

Put a command Button on a Form and do....

Sub Command1_Click
Me.Mousepointer = vbhourglass
..some sort of big loop to make a delay
end sub

The mousepoiner does NOT change till the delay loop has executed.
Me.mousepointer and Screen.Mousepointer have similar issue.
Me.Refresh or DoEvents do NOT fix this issue.

**Expected behavior**
Mousepointer should change instantly.
Setting VBhourglass before a slow bit of code is what its all about, so having no response till after the slow code has executed is a real problem

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 11]
 - twinBASIC compiler version [168]

**Additional context**

**Comment by AntsAntix** [comment]
Surely, this is all about interrupts. I you write tight, processor-bound code you will not achieve what you look for.

**Comment by dmrvb** [comment]
VB doesn't allow an ordinary user to handle interupts, its very much single threaded. Do you mean message handling?
The mousepointer and many other GUI features respond instantly in VB and tB aims to be 100% compatible.
Setting the mousepointer to Hourglass before starting a bit of processor intensive code is a very common requirement.
Having full control over the priority of user code and GUI operation would be a good feature but thats for the future. 
Note that DoEvents does not fix the current problem.

**Comment by fafalone** [comment]
twinBASIC very much supports multithreading though. Right now, not in the future. In the future there will be built in syntax... but right now, you [can use](https://github.com/fafalone/ThreadTest) all the multithreading APIs.

**Comment by dmrvb** [comment]
Ages ago I had a very quick look at multithreading with API calls in VB6 and rapidly gave up. If tB has "easy" multi-threading in the future then that will be great.
Instant updating of the mousepointer is how VB6 works so tB should do the same to be compatible.

**Comment by AntsAntix** [comment]
@dmrvb - my hardware knowledge is really rusty . . . but at a really low level, control of the CPU is governed by interrupts and this will be at a Windows kernel level. If your generated code takes control of the CPU and does not allow Windows to interrupt and split processor 'responsibilty', you may not _**see**_ the response you look for. And, yes, it might well reflect in Windows messaging system ... 
But my response was more to suggest that perhaps the code you have used in the loop is such that Windows does not get the opportunity to inject a message to bring about the change in MousePointer.

**Comment by fafalone** [comment]
Yeah because of the runtime it's *very* difficult to use them in VB6; it will crash if you do *anything* the involves the runtime, which is virtually everything. You need elaborate hacks... which exist, The trick has a module that you can just drop in and use like CreateThread.

But since tB lacks a runtime, there's no crash when you call CreateThread. My Event Tracer for a more complex example runs two threads without any hacks.

Right now tB has a bug introduced in Beta 146 that's causing some minor issues with it though (code reordering is messing up some things, though it still works without crashing), but I'm sure that will be fixed now that I've found it.

**Comment by fafalone** [comment]
Actually this call reordering issue sounds like it might be causing the mousepointer problem too.

**Comment by WaynePhillipsEA** [comment]
I think this is just a normal bug, and easily fixed.  Currently the mouse pointer will only get updated in response to WM_SETCURSOR events, but those events won't fire until your procedure completes.  tB just needs to update it immediately once set.

**Comment by dmrvb** [comment]
> @dmrvb - my hardware knowledge is really rusty . . . but at a really low level, control of the CPU is governed by interrupts and this will be at a Windows kernel level. If your generated code takes control of the CPU and does not allow Windows to interrupt and split processor 'responsibilty', you may not _**see**_ the response you look for. And, yes, it might well reflect in Windows messaging system ... But my response was more to suggest that perhaps the code you have used in the loop is such that Windows does not get the opportunity to inject a message to bring about the change in MousePointer.

In a previous life I have written interrupt handling software on a mini-computer, and more recently on the Arduino, but when it comes to Windows I think we have to let go of this "hardware thinking", there is so much clever mostly software based stuff going on and there is multi-threading even if VB does not let us see it. Windows can easily take priority over any processor intensive loop that we can create. Note how Windows changes your window title to "not responding". Thats not even your window, its a snapshot of your window taken by Windows.  :-)

**Comment by fafalone** [comment]
Why let go of hardware thinking when tB supports kernel mode programming and driver development? :)

It's an exciting time for opening up lower level programming to easier tools.

**Comment by dmrvb** [comment]
Well yes, tB should allow some clever low level stuff, but for most users VB was an easy to learn (?) high level language. My real point was that for 99% of users Windows totally isolates us from the hardware and we have to think in terms of abstract windows concepts.  I still half think of registers and 1 and 0's when I program, but I don't think thats typical, Im an engineer who does some programming   Writing something in tB to actually capture and handle an interrupt would be an interesting challenge. 
I do suspect that the people here probably over-represent the real nuts and bolts experts rather than the typical high level VB6 user, but thats probably a good thing.

**Comment by rexxitall** [comment]
As long as the nuts and bolts fraction do not forget who pays their bills all is fine. :) Its nice to have high level concepts  but it is bad if its stops there. Then you are in the area of nasty work arounds to get the things the high level hides. :) And it is aso tricky how to show it in a typically BASIC way. Especially if they are new to basic. But thats why they have github and discord to discuss it :) And seems they are very patient :)

**Comment by dmrvb** [comment]
It would be great if VB/tB could be a self contained high level language, but that would likely have huge limitations. I want to use VB as a high level language but have done all my graphics in GDI and all my audiio with (some quite difficult) API code to get speed and flexibility. There is a lot of discussion here about the nuts and bolts of COM, I don't want to have to understand this, but VB/tB are built on COM so its important to get it right, its providing the foundations for our high level stuff. Its good there are people here who have the required knowledge.
There are a number of almost day to day tasks that can't be done without API calls or subclassing and it would good to resolve these for those who want to be 100% high level.

**Comment by fafalone** [comment]
What I think it great about VB is it has easy to understand syntax and does high level RAD like nothing else ever made, but still lets you pretty easily drop down to the lower levels as you need them.

Languages usually pick one or the other. .NET is awful for low level; it's impossible to do kernel mode it it and all the others like it.

**Comment by WaynePhillipsEA** [comment]
Fixed in BETA 169.  Thanks for reporting!
